staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/buffers.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package modes
+
+import (
+	"bytes"
+	"sync"
+)
+
+var buffers BufferProvider
+
+type buffer struct {
+	bytes.Buffer
+}
+
+type BufferProvider struct {
+	p sync.Pool
+}
+
+func (b *BufferProvider) Get() *buffer {
```

Comment:
Optimizing this in the future might be possible if somehow the Get caller could provide a size hint... Is that at all possible? (I didn't see anything in this PR that would help, but I'm not as familiar with the CBOR library, so figured I'd ask)

Reply:
It can definitely be optimized in the future. The library currently either uses an internal `sync.Pool` of `byte.Buffer` or accepts a buffer from the user on each marshal call. We can come up with the buffer however we like, so this BufferProvider contract can change to accept a size hint, or itself select a buffer capacity based on the size distribution of recently encoded objects of the same type, etc.

That said, I think it will turn out to be more impactful to make sure that list encoding can be streamed.